src/index.ts

```diff
@@ -0,0 +1,641 @@
+	if ( Error.captureStackTrace ) {
+	    Error.captureStackTrace( this, HoloError );
+	}
+
+	this.name			= 'HoloError';
+
+	// Fix for Typescript
+	//   - https://github.com/Microsoft/TypeScript/wiki/Breaking-Changes#extending-built-ins-like-error-array-and-map-may-no-longer-work
+	Object.setPrototypeOf( this, HoloError.prototype );
+    }
+
+    toJSON () {
+	return {
+	    "name": this.name,
+	    "message": this.message,
+	};
+    }
+}
+
+
+class Envoy {
+    ws_server		: any;
+    http_server		: any;
+    opts		: any;
+    conductor_opts	: any;
+    connected		: any;
+
+    request_counter	: number	= 0;
+    entry_counter	: number	= 0;
+    pending_confirms	: object	= {};
+    pending_entries	: object	= {};
+    anonymous_agents	: any		= {};
+
+    hcc_clients		: any		= {};
+
+    constructor ( opts ) {
+	this.opts			= Object.assign({}, {
+	    "port": WS_SERVER_PORT
+	}, opts);
+
+	this.conductor_opts		= {
+	    "interfaces": {
+		"master_port":		42211,
+		"service_port":		42222,
+		"internal_port":	42233,
+		"public_port":		42244,
+	    },
+	};
+
+	this.connections();
+	this.startWebsocketServer();
+	this.startHTTPServer();
+    }
+
+    async connections () {
+	try {
+	    const ifaces		= this.conductor_opts.interfaces;
+
+	    this.hcc_clients.master	= new WebSocket(`ws://localhost:${ifaces.master_port}`,   RPC_CLIENT_OPTS );
+	    this.hcc_clients.service	= new WebSocket(`ws://localhost:${ifaces.service_port}`,  RPC_CLIENT_OPTS );
+	    this.hcc_clients.internal	= new WebSocket(`ws://localhost:${ifaces.internal_port}`, RPC_CLIENT_OPTS );
+	    this.hcc_clients.general	= new WebSocket(`ws://localhost:${ifaces.public_port}`,   RPC_CLIENT_OPTS );
+	} catch ( err ) {
+	    console.error( err );
+	}
+
+	const clients			= Object.values( this.hcc_clients );
+	this.connected			= Promise.all( clients.map( (client:any) => client.opened( CONDUCTOR_TIMEOUT ) ))
+    }
+
+    async startWebsocketServer () {
+	this.ws_server			= new WebSocketServer({
+	    "port": this.opts.port,
+	    "host": "localhost",
+	});
+
+	this.ws_server.on("connection", async (socket, request) => {
+	    // path should contain the HHA ID and Agent ID so we can do some checks and alert the
+	    // client-side if something is not right.
+	    log.info("URL: %s", request.url );
+	    const url			= new URL( request.url, "http://localhost");
+
+	    const anonymous		= url.searchParams.get('anonymous') === "true" ? true : false;
+	    const agent_id		= url.searchParams.get('agent_id');
+	    const hha_hash		= url.searchParams.get('hha_hash');
+
+	    if ( anonymous ) {
+		log.debug("Adding Agent %s to anonymous list for hApp %s", agent_id, hha_hash );
+		this.anonymous_agents[ agent_id ]	= hha_hash;
+
+		socket.on("close", async () => {
+		    log.debug("Remove Agent %s from anonymous list", agent_id, hha_hash );
+		    delete this.anonymous_agents[ agent_id ];
+		});
+	    }
+	});
+
+	this.ws_server.register("holo/agent/identify", async ([ agent_id ]) => {
```

Comment:
We should demand a Signature from the client, proving that they are indeed holding the private key matching this public key (Agent ID).  Otherwise, randos can claim to be an Agent, and start running Zome API calls (ie. that read and return private data) on this hosted agent.

Actually, this isn't a security hole.  And, every Zome API call, invokes servicelogger's `log_request` API first thing, which *verifies* that the Client actually signed call `request` package.  It will fail invalid requests immediately.

Reply:
However, we should use a signature of Iso8601 the current time ("YYYY-MM-DDTHH:MM:00Z", rounded down to the current minute), and sign it.  This signature should be passed (I think to the "connection" endpoint, which occurs in earliest real-time).  It should verify the signature, against the current and previous Iso8601 Zulu minute, and only allow the connection.  If possible, we can return the HTTP error response (wrong timestamp signature).

Or, return a blog of entropy (a challenge) from the connection, that has to be signed.  Or whatever.